Always set REMOTE_USER when valid auth provided (even if not required).

twosigmajab opened the pull request:
Interpret "auth not required" (i.e. `auth_required_callback` returning False)
to mean: missing/invalid auth will not result in a 40x error,
but if a client does provide valid auth for an auth-not-required request,
the remote user will be exposed to the app, since it can still be useful to know .

This better supports cases where auth is not required for a particular request
(e.g. a health check endpoint hit anonymously by Kubernetes),
but it is still useful to know who made the request when valid auth is provided
(e.g. a script or user interactively hitting the health check endpoint
from a client that does pre-emptive Kerberos -- in this case the app
can check if the remote user is an admin, and display richer info
in the health check response body).

Includes unit tests ~and also updates the "Changes" section of the README~ (UPDATE: removed release notes changes originally added to the README, as requested in a different PR).

twosigmajab committed:
Validate rather than ignore first word in header.

Validate that the first word is "Negotiate" rather than ignoring it.

Fixes #16.

twosigmajab commented:
I based this off my branch with the fix for #16, since I wanted to incorporate that fix in this revision. If this PR is accepted, GitHub should automatically resolve #17 and #16 along with it.

twosigmajab commented:
Thanks for reviewing, @hawicz!

propertone commented on wsgi_kerberos.py:
```diff
@@ -172,19 +172,32 @@ def __call__(self, environ, start_response):
+                error_resp(environ, start_response)
+                if auth_required
+                else self.application(environ, start_response)
+            )
```
Can you rewrite this as a regular if/else block? It's more readable and whitespace isn't expensive.

twosigmajab replied:
Done in the latest revision.

propertone commented on wsgi_kerberos.py:
```diff
@@ -202,7 +215,7 @@ def custom_start_response(status, headers, exc_info=None):
         elif server_token:
             # If we got a token, but no user, return a 401 with the token
-            return self._unauthorized(environ, start_response, server_token)
+            return _40x_resp_if_auth_required(self._unauthorized)
```
I think this is incorrect - if auth is not required, but the client sends an authorization header that doesn't include a user (ie: to verify mutual authentication), then it should send back a 401 with a server token in any case (@mkomitee correct me if I'm wrong).

twosigmajab replied:
UPDATE: Sorry I initially misunderstood what you were saying. This comment should be ignored.

~This gave me pause too, but based on https://tools.ietf.org/html/rfc4559#page-4 I think it's correct? (Note the 403 behavior in this `elif` block is not introduced by this PR, it is pre-existing.)~

> ~If a 401 containing a "WWW-Authenticate" header with "Negotiate" and gssapi-data is returned from the server, it is a continuation of the authentication request.~

~Based on the RFC and the code's interpretation, I believe 401 means "please try (or continue trying) to authenticate", whereas 403 is overloaded and means either "you authenticated but are not authorized", or, "you tried to authenticate, it failed, and you should not continue trying". Responding with another 401 in that latter case would indicate you want the client to keep trying, which you don't.~

twosigmajab replied:
UPDATE: Sorry I initially misunderstood what you were saying. This comment should be ignored.

~So 403 rather than 401 is preferable for the "Negotiate INVALID" case to stop the client from continuing GSSAPI negotiation. Does that make sense?~

propertone replied:
The issue here is the `server_token` variable is no longer passed to `self._unauthorized` if auth is required. Multi-round trip negotiate cycle is technically a normal part of the GSS-API security context establishment spec (though not much used in practice). From https://www.researchgate.net/figure/SIP-REGISTER-message-flow-with-GSS-API-security-context-establishment-authentication_fig4_230585165 :

```
...Depending on the underlying security mechanism, additional token exchanges may be required in the course of context establishment. If so, GSS_S_CONTINUE_NEEDED status is set and additional tokens are passed between the client and server until a security context is established,...
```

twosigmajab replied:
Sorry I initially misunderstood what you were saying. Thanks for the good catch; should be fixed in the latest revision. Another look?

twosigmajab committed:
Always set REMOTE_USER when valid auth provided (even if not required).

Interpret "auth not required" (i.e. auth_required_callback returning False)
to mean: missing/invalid auth will not result in a 40x error,
but if a client provides valid auth, it will now be exposed to the app.

This better supports cases where auth is not required for a particular request
(e.g. a health check endpoint hit anonymously by Kubernetes),
but it is still useful to know who made the request when valid auth is provided
(e.g. a script or user interactively hitting the health check endpoint
from a client that does pre-emptive Kerberos).

propertone reviewed (approved):
Looks good
